docs/use_cases/hybrid_search_&_rerank_rag.md

```diff
@@ -0,0 +1,314 @@
+
+```
+Create a vector store using the embeddings we obtain from the text.
+
+```python
+embeddings = HuggingFaceInferenceAPIEmbeddings(
+    api_key=HF_TOKEN, model_name="BAAI/bge-base-en-v1.5"
+)
+vectorstore = Chroma.from_documents(chunks, embeddings)
+```
+Now build the keyword and semantic retrievers separately. For keyword matching, we use the [BM25 retriever](https://python.langchain.com/docs/integrations/retrievers/bm25) from Langchain. By setting k to 3, you’re asking the retriever to return the 3 most relevant documents or vectors from the vector store.
+
+```python
+vectorstore_retreiver = vectorstore.as_retriever(search_kwargs={"k": 3})
+keyword_retriever = BM25Retriever.from_documents(chunks)
+keyword_retriever.k =  3
+```
+
+Now create the ensemble retriever which is a weighted combination of the above two.
+
+```python
+ensemble_retriever = EnsembleRetriever(retrievers=[vectorstore_retreiver,
+                                                   keyword_retriever],
+                                       weights=[0.3, 0.7])
+```
+Modify the weights to incorporate the impact of both search outcomes appropriately.  The values correspond to **α** and **1-α** as we discussed above. Here I have weighed keywords more with a value of 0.7.
+
+We need to use an LLM in our RAG pipeline. Here we utilize a quantized version of [Zephyr-7B-Beta](http://HuggingFaceH4/zephyr-7b-beta) for lightweight and optimized performance.
+
+```python
+model_name = "HuggingFaceH4/zephyr-7b-beta"
+
+# function for loading 4-bit quantized model
+def load_quantized_model(model_name: str):
+    """
+    model_name: Name or path of the model to be loaded.
+    return: Loaded quantized model.
+    """
+    bnb_config = BitsAndBytesConfig(
+        load_in_4bit=True,
+        bnb_4bit_use_double_quant=True,
+        bnb_4bit_quant_type="nf4",
+        bnb_4bit_compute_dtype=torch.bfloat16,
+    )
+
+    model = AutoModelForCausalLM.from_pretrained(
+        model_name,
+        load_in_4bit=True,
+        torch_dtype=torch.bfloat16,
+        quantization_config=bnb_config,
+    )
+    return model
+
+# initializing tokenizer
+def initialize_tokenizer(model_name: str):
+    """
+    model_name: Name or path of the model for tokenizer initialization.
+    return: Initialized tokenizer.
+    """
+    tokenizer = AutoTokenizer.from_pretrained(model_name, return_token_type_ids=False)
+    tokenizer.bos_token_id = 1  # Set beginning of sentence token id
+    return tokenizer
+
+tokenizer = initialize_tokenizer(model_name)
+
+model = load_quantized_model(model_name)
+# specify stop token ids
+stop_token_ids = [0]
+
+# build huggingface pipeline for using zephyr-7b-beta
+pipeline = pipeline(
+    "text-generation",
+    model=model,
+    tokenizer=tokenizer,
+    use_cache=True,
+    device_map="auto",
+    max_length=2048,
+    do_sample=True,
+    top_k=5,
+    num_return_sequences=1,
+    eos_token_id=tokenizer.eos_token_id,
+    pad_token_id=tokenizer.eos_token_id,

```

Comment:
are we padding with the eos token? shouldn't this be the pad_token?
for reference: https://huggingface.co/docs/transformers/main_classes/tokenizer

Reply:
Yes, my mistake. That is ".pad_token_id."